cargo/cargo_build_script_runner/lib.rs

```diff
@@ -0,0 +1,159 @@
+// Copyright 2018 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Parse the output of a cargo build.rs script and generate a list of flags and
+//! environment variable for the build.
+use std::io::{BufRead, BufReader, Read};
+use std::process::{Command, Stdio};
+
+/// Enum containing all the considered return value from the script
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum BuildScriptOutput {
+    /// Value for ignored cargo directive
+    None,
+    /// cargo:rustc-link-lib
+    LinkLib(String),
+    /// cargo:rustc-link-search
+    LinkSearch(String),
+    /// cargo:rustc-cfg
+    Cfg(String),
+    /// cargo:rustc-flags
+    Flags(String),
+    /// cargo:rustc-env
+    Env(String),
+}
+
+impl BuildScriptOutput {
+    /// Converts a line into a [BuildScriptOutput] enum.
+    ///
+    /// Examples
+    /// ```rust
+    /// assert_eq!(BuildScriptOutput::new("cargo:rustc-link-lib=lib"), BuildScriptOutput::LinkLib("lib".to_owned()));
+    /// ```
+    fn new(line: &str) -> BuildScriptOutput {
+        let split = line.splitn(2, '=').collect::<Vec<_>>();
+        if split.len() <= 1 {
+            return BuildScriptOutput::None;
+        }
+        let param = split[1].trim().to_owned();
+        match split[0] {
+            "cargo:rustc-link-lib" => BuildScriptOutput::LinkLib(param),
+            "cargo:rustc-link-search" => BuildScriptOutput::LinkSearch(param),
+            "cargo:rustc-cfg" => BuildScriptOutput::Cfg(param),
+            "cargo:rustc-flags" => BuildScriptOutput::Flags(param),
+            "cargo:rustc-env" => BuildScriptOutput::Env(param),
+            _ => BuildScriptOutput::None,
```

Comment:
What are your thoughts on at least printing a warning in this case?

The following look (currently) unsupported, and I imagine build script syntax may be expanded in the future.

- rerun-if-changed
- rerun-if-env-changed
- rustc-cdynlib-link-arg
- warning
- KEY=VALUE (for links scripts)

(from https://doc.rust-lang.org/cargo/reference/build-scripts.html#outputs-of-the-build-script)

Reply:
Done. (although explicitely ignored rerun*)